### PR TITLE
server: Fix some cpuspeed issues while create service offering

### PR DESCRIPTION
**Pull request by ustcweizhou**
## Description

Only the following offerings are accepted

(1) fixed offering:
cpunumber, memory, must be specified.
cpuspeed should be specified

example: create serviceoffering displaytext=offering-fix name=offering-fix cpunumber=1 memory=1024 cpuspeed=1000

(should not work)
create serviceoffering displaytext=offering-fix name=offering-fix cpunumber=1 memory=1024 (but works in 4.14, fixed by this commit)

(2) unconstrained offering:

cpunumber, memory must be null
cpuspeed must be null

example: create serviceoffering displaytext=offering-unconstrained name=offering-unconstrained

(should not work)
create serviceoffering displaytext=offering-unconstrained name=offering-unconstrained cpunumber=1
create serviceoffering displaytext=offering-unconstrained name=offering-unconstrained memory=1024
create serviceoffering displaytext=offering-unconstrained name=offering-unconstrained cpuspeed=1000 (but works in 4.14, fixed by this commit)

(3) constrained offering:

cpunumber, memory must be null
mincpunumber, maxcpunumber, minmemory, maxmemory, cpuspeed must be specified

example:create serviceoffering displaytext=offering-constrained name=offering-constrained mincpunumber=1 maxcpunumber=2 minmemory=1024 maxmemory=2048 cpuspeed=1000

(should not work)
create serviceoffering displaytext=offering-constrained name=offering-constrained mincpunumber=1 maxcpunumber=2 minmemory=1024 maxmemory=2048 (but works in 4.14, fixed by this commit)

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Review by DaanHoogland — APPROVED** [review]
code lgtm

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by shwstppr — APPROVED** [review]
LGTM

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 1]
```diff
@@ -2319,9 +2319,9 @@ public ServiceOffering createServiceOffering(final CreateServiceOfferingCmd cmd)
...
             // if any of them is null, then all of them shoull be null
-            if (maxCPU == null || minCPU == null || maxMemory == null || minMemory == null) {
-                if (maxCPU != null || minCPU != null || maxMemory != null || minMemory != null) {
-                    throw new InvalidParameterValueException("For creating a custom compute offering min/max cpu and min/max memory should all be specified");
+            if (maxCPU == null || minCPU == null || maxMemory == null || minMemory == null || cpuSpeed == null) {
+                if (maxCPU != null || minCPU != null || maxMemory != null || minMemory != null || cpuSpeed != null) {
+                    throw new InvalidParameterValueException("For creating a custom compute offering min/max cpu and min/max memory/cpu speed should all be null or all specified");
                 }
             } else {
                 if (cpuSpeed != null && (cpuSpeed.intValue() < 0 || cpuSpeed.longValue() > Integer.MAX_VALUE)) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:2327]: `cpuSpeed` null check not required here, it is always not null as per the above changes.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -2343,7 +2343,7 @@ public ServiceOffering createServiceOffering(final CreateServiceOfferingCmd cmd)
             if (cpuNumber != null && (cpuNumber.intValue() <= 0 || cpuNumber.longValue() > Integer.MAX_VALUE)) {
                 throw new InvalidParameterValueException("Failed to create service offering " + offeringName + ": specify the cpu number value between 1 and " + Integer.MAX_VALUE);
             }
-            if (cpuSpeed != null && (cpuSpeed.intValue() < 0 || cpuSpeed.longValue() > Integer.MAX_VALUE)) {
+            if (cpuSpeed == null || (cpuSpeed.intValue() < 0 || cpuSpeed.longValue() > Integer.MAX_VALUE)) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:2346]: @ustcweizhou similar cond check applicable for `cpuNumber` and `memory` if these are mandatory params? check and update accordingly.

Comment (weizhouapache) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:2346]: @sureshanaparti cpuNumber and memory are not mandatory for dynamic offerings.
There are 3 types of offerings
(1) fixed offering, cpu number, cpu speed, memory are mandatory
(2) dynamic offering, 
(2.1) unconstrained offering, cpu number, cpu speed, memory are not required
(2.2) constrained offerings, min/max cpu, min/max memory, and cpu speed are mandatory.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 1 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:2327]: @sureshanaparti indeed, I will remove it.

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java (hunk 2 above)
Comment (sureshanaparti) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:2346]: @weizhouapache thanks for confirming

**Review by sureshanaparti — APPROVED** [review]

**Pull request merged**